File: export-templates/javascript/src/components/examples.js
const { OpenAIEmbeddings } = require('@langchain/openai');
const { MemoryVectorStore } = require('langchain/vectorstores/memory');
const fs = require('fs');
const path = require('path');
const { IS_LAMBDA, getOpenAIApiKey } = require('../constants');

let embeddings = null;
let vectorStore = null;
let initialized = false;
let documents = null;

function decodeBase64Embedding(b64String) {
  const buffer = Buffer.from(b64String, 'base64');
  const decodedArray = new Float32Array(buffer);
  return decodedArray;
}

async function initializeExamples() {
  const apiKey = await getOpenAIApiKey();

  embeddings = new OpenAIEmbeddings({
    model: process.env.OPENAI_EMBEDDING_MODEL || 'text-embedding-3-small',
    apiKey: apiKey
  });

  const examplesDirectory = IS_LAMBDA ? path.resolve('/var/task/examples') : path.join(__dirname, '../examples');
  const examples = fs.readdirSync(examplesDirectory)
    .filter(file => file.endsWith('.json'))
    .map(file => {
      const content = fs.readFileSync(path.join(examplesDirectory, file), 'utf8');
      return JSON.parse(content);
    });

  documents = examples.map(example => 
    ({ pageContent: JSON.stringify({ ...example.document, embedding: null }), metadata: {} })
  );
  const embeddingsList = examples.map(example => 
    decodeBase64Embedding(example.taskEmbedding)
  );
  vectorStore = new MemoryVectorStore();
  vectorStore.addVectors(embeddingsList, documents);
  initialized = true;
}

async function getEmbedding(text) {
  if (!initialized) throw new Error('examples.js not initialized. Call initializeExamples() first.');
  
  text = text.replace("\n", " ")
  return await embeddings.embedQuery(text)
}

async function getRelevantExamples(query, k = 10) {
  if (!initialized) throw new Error('examples.js not initialized. Call initializeExamples() first.');
  
  const queryEmbedding = await getEmbedding(query);
  const results = await vectorStore.similaritySearchVectorWithScore(queryEmbedding, k);

  const relevantExamples = [];
  for (const doc of results) {
    const example = JSON.parse(doc[0].pageContent);
    relevantExamples.push({
      task: example.task,
      code: example.code
    });
  }
  return relevantExamples;
}

module.exports = {
  initializeExamples,
  getRelevantExamples
};

File: examples/bank-report-demo-py/export/src/components/examples.py
import os
import json
import base64
import numpy as np
from langchain_openai import OpenAIEmbeddings
from typing import List, Dict
from langchain_community.vectorstores import FAISS
from constants import IS_LAMBDA

# Initialize embeddings
embeddings = OpenAIEmbeddings(model="text-embedding-ada-002")

def decode_base64_embedding(b64_string):
    decoded_bytes = base64.b64decode(b64_string)
    return np.frombuffer(decoded_bytes, dtype=np.float32)

examples_directory = os.path.join('/var/task/examples' if IS_LAMBDA else os.path.dirname(__file__), '..', 'examples')
examples_with_embeddings = []

for filename in os.listdir(examples_directory):
    if filename.endswith('.json'):
        with open(os.path.join(examples_directory, filename), 'r') as f:
            examples_with_embeddings.append(json.load(f))

# Create vector store from pre-computed embeddings
texts = [json.dumps({**example['document'], 'embedding': None}) for example in examples_with_embeddings]
embeddings_list = [decode_base64_embedding(example['taskEmbedding']) for example in examples_with_embeddings]
metadatas = [{}] * len(examples_with_embeddings)

# Convert list of numpy arrays to 2D numpy array
embeddings_array = np.array(embeddings_list)

vector_store = FAISS.from_embeddings(
    text_embeddings=list(zip(texts, embeddings_list)),
    embedding=embeddings,
    metadatas=metadatas
)

def get_embedding(text):
    text = text.replace("\n", " ")
    return embeddings.embed_query(text)

def get_relevant_examples(query: str, k: int = 10) -> List[Dict]:
    # Generate embedding for the query using the same model as stored embeddings
    query_embedding = get_embedding(query)
 
    # Perform vector search
    results = vector_store.similarity_search_by_vector(query_embedding, k=k)

    relevant_examples = []
    for doc in results:
        example = json.loads(doc.page_content)
        relevant_examples.append({
            "task": example["task"],
            "code": example["code"]
        })
    
    return relevant_examples